trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/TestAuthzBeforeRewrite.java

```diff
@@ -0,0 +1,132 @@
+import io.trino.aws.proxy.server.testing.TestingTrinoAwsProxyServer;
+import io.trino.aws.proxy.server.testing.containers.S3Container.ForS3Container;
+import io.trino.aws.proxy.server.testing.harness.BuilderFilter;
+import io.trino.aws.proxy.server.testing.harness.TrinoAwsProxyTest;
+import io.trino.aws.proxy.server.testing.harness.TrinoAwsProxyTestCommonModules.WithConfiguredBuckets;
+import io.trino.aws.proxy.spi.credentials.Credentials;
+import io.trino.aws.proxy.spi.security.SecurityResponse;
+import io.trino.aws.proxy.spi.security.opa.OpaS3SecurityFacade;
+import org.assertj.core.api.InstanceOfAssertFactories;
+import org.junit.jupiter.api.Test;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+
+import java.net.URI;
+import java.util.Optional;
+
+import static com.google.inject.multibindings.OptionalBinder.newOptionalBinder;
+import static io.trino.aws.proxy.server.security.opa.OpaS3SecurityModule.OPA_S3_SECURITY_IDENTIFIER;
+import static io.trino.aws.proxy.server.testing.RequestRewriteUtil.CREDENTIAL_TO_REDIRECT;
+import static io.trino.aws.proxy.server.testing.RequestRewriteUtil.TEST_CREDENTIAL_REDIRECT_BUCKET;
+import static io.trino.aws.proxy.server.testing.RequestRewriteUtil.TEST_CREDENTIAL_REDIRECT_KEY;
+import static io.trino.aws.proxy.server.testing.TestingUtil.TEST_FILE;
+import static io.trino.aws.proxy.server.testing.TestingUtil.assertFileNotInS3;
+import static io.trino.aws.proxy.server.testing.TestingUtil.clientBuilder;
+import static io.trino.aws.proxy.spi.plugin.TrinoAwsProxyServerBinding.s3RequestRewriterModule;
+import static java.util.Objects.requireNonNull;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@TrinoAwsProxyTest(filters = {WithConfiguredBuckets.class, TestAuthzBeforeRewrite.Filter.class})
+public class TestAuthzBeforeRewrite
+{
+    final S3Client s3Client;
+    final S3Client storageClient;
+    private final TestingS3RequestRewriter s3requestRewriteController;
+
+    private final URI baseUri;
+    private final String relativePath;
+
+    public static class Filter
+            implements BuilderFilter
+    {
+        @Override
+        public TestingTrinoAwsProxyServer.Builder filter(TestingTrinoAwsProxyServer.Builder builder)
+        {
+            return builder.withProperty("s3-security.type", OPA_S3_SECURITY_IDENTIFIER)
+                    .withProperty("opa-s3-security.server-base-uri", "http://localhost/v1/data")
+                    .addModule(binder -> binder.bind(OpaS3SecurityFacade.class).toInstance((_, _, _, _) -> SecurityResponse.FAILURE))
+                    .addModule(s3RequestRewriterModule("testing", TestingS3RequestRewriter.class, binder -> {
```

Comment:
We have

```java
RequestRewriteUtil.Filter
```

that does something similar, could we reuse that?

it doesn't count call but you can add that